bugsnag-android-core/src/main/java/com/bugsnag/android/DocumentPath.java

```diff
@@ -0,0 +1,385 @@
+    }
+
+    private static boolean isDotLast(String path) {
+        if (path.charAt(path.length() - 1) != '.') {
+            return false;
+        }
+
+        if (path.length() == 1) {
+            return true;
+        }
+
+        boolean isUnescapedDot = true;
+        for (int i = path.length() - 2; i >= 0; i--) {
+            char ch = path.charAt(i);
+            if (ch == '\\') {
+                isUnescapedDot = !isUnescapedDot;
+            } else {
+                break;
+            }
+        }
+
+        return isUnescapedDot;
+    }
+
+    private static Directive makeDirectiveFromPathComponent(String pathComponent) {
+        if (isInteger(pathComponent)) {
+            int index = Integer.parseInt(pathComponent);
+            if (index == -1) {
+                return new Directive.ListLastIndexDirective();
+            } else {
+                return new Directive.ListIndexDirective(index);
+            }
+        }
+        return new Directive.MapKeyDirective(pathComponent);
+    }
+
+    private static boolean isInteger(String str) {
+        if (str.length() == 0) {
+            return false;
+        }
+
+        int startIndex = 0;
+        if (str.charAt(0) == '-') {
+            startIndex = 1;
+        }
+
+        for (int i = startIndex; i < str.length(); i++) {
+            int ch = str.charAt(i);
+            if (ch < '0' || ch > '9') {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * A directive determines how document path commands are interpreted. Depending on the
+     * current context, making a container, getting from a container, or setting to a container
+     * will require different operations to complete.
+     */
+    private interface Directive {
+        /**
+         * Make a new container at the current level.
+         *
+         * @return The new container.
+         */
+        @NonNull Object newContainer();
+
+        /**
+         * Get an object from the container at the current level.
+         *
+         * @param container The container to get the object from.
+         * @return The resulting object, or null if not found.
+         */
+        @Nullable Object getFromContainer(@NonNull Object container);
+
+        /**
+         * Set an object in a container at the current level.
+         *
+         * @param container The container to set a value in.
+         * @param value The value to set (can be null, which usually means "remove")
+         */
+        void setInContainer(@NonNull Object container, @Nullable Object value);
+
+        /**
+         * Modifies a particular key in a map, or creates a new (String, Object) map.
+         * Setting null removes the value.
+         */
+        class MapKeyDirective implements Directive {
+            private final String key;
+
+            public MapKeyDirective(@NotNull String key) {
+                this.key = key;
+            }
+
+            @NotNull
```

Comment:
`@NonNull` is the preferred annotation (but frustratingly android studio sometimes picks `@NotNull`)